### PR TITLE
Feat/blockchain properties

### PR DESCRIPTION
**Pull request by patryk-kochanski**

### CODE REVIEW

File: packages/origin-247-certificate/migrations/1646865380689-Add deploymentProperties.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddDeploymentProperties1646865380689 implements MigrationInterface {
+    name = 'AddDeploymentProperties1646865380689';
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(
+            `CREATE TABLE "deployment_properties" ("registry" character varying NOT NULL, CONSTRAINT "PK_0b3f711854067e456735f0d25eb" PRIMARY KEY ("registry"))`
```
Comment (soanvig) [inline packages/origin-247-certificate/migrations/1646865380689-Add deploymentProperties.ts:8]: I think this should be prefixed with standard prefix for 247-certificate tables (`certificate_` afair)

File: packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    constructor(
+        private configService: ConfigService,
+        private deploymentPropsRepo: DeploymentPropertiesRepository
+    ) {}
+
+    async getProperties(): Promise<IBlockchainProperties> {
+        const primaryRPC = this.configService.get<string>('WEB3');
+        const issuerKey = this.configService.get<string>('ISSUER_KEY');
+
+        if (!primaryRPC) {
```
Comment (soanvig) [inline packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts:19]: I would validate that in constructor

File: packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import { Injectable, OnModuleInit } from '@nestjs/common';
+import { IBlockchainProperties, Contracts } from '@energyweb/issuer';
+import { ConfigService } from '@nestjs/config';
+import { DeploymentPropertiesRepository } from './repositories/deploymentProperties/deploymentProperties.repository';
+import { providers, Signer, Wallet } from 'ethers';
+import { getProviderWithFallback } from '@energyweb/utils-general';
+
+@Injectable()
+export class BlockchainPropertiesService {
```
Comment (soanvig) [inline packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts:9]: I think you wanted to change this name?

**Review by soanvig — CHANGES_REQUESTED** [review]
README is missing

File: packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts (hunk 2 above)
Comment (patryk-kochanski) [reply packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts:22]: Done

File: packages/origin-247-certificate/README.md [hunk 4]
```diff
@@ -88,6 +88,22 @@ export class SomeService {
 }
 ```
 
+## Deploying and migrating contracts
+
+It is possible to deploy contracts using `OnChainCertificateFacade`. When using this method, it is unnecesary to do blockchain seeding migrations. How to use:
+
+-   import certificate module into one of your applictaions modules
```
Comment (soanvig) [inline packages/origin-247-certificate/README.md:95]: typo :byakugan:

File: packages/origin-247-certificate/README.md [hunk 5]
```diff
@@ -88,6 +88,22 @@ export class SomeService {
...
+-   at application bootstrap, use OnChainCertificateFacade.deploy() method
```
Comment (soanvig) [inline packages/origin-247-certificate/README.md:96]: Add backticks to wrap code

File: packages/origin-247-certificate/README.md [hunk 6]
```diff
@@ -88,6 +88,22 @@ export class SomeService {
...
+
+```ts
+import { OnChainCertificateFacade } from '@energyweb/origin-247-certificate';
+// ...
+const appModule = AppModule.register();
+const app = await NestFactory.create(appModule);
+const blockchainFacade = await app.resolve<OnChainCertificateFacade>(OnChainCertificateFacade);
+await blockchainFacade.deploy();
```
Comment (soanvig) [inline packages/origin-247-certificate/README.md:104]: Just make it ready to copy paste:

```
constructor(
   private onChangeCertificateFacade: OnChainCertificateFacade
) {}

public async onApplicationBootstrap() {
   await this.onChangeCertificateFacade.deploy();
}
```

File: packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        this.primaryRPC = primaryRPC;
+        this.issuerPrivateKey = issuerPrivateKey;
+    }
+
+    async getProperties(): Promise<IBlockchainProperties> {
+        const { registry } = await this.deploymentPropsRepo.get();
+
+        const web3 = getProviderWithFallback(...[this.primaryRPC].filter((url) => Boolean(url)));
+
+        let signer: Signer = new Wallet(this.assure0x(this.issuerPrivateKey), web3);
```
Comment (soanvig) [inline packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts:37]: const ?

File: packages/origin-247-certificate/test/setup.ts [hunk 8]
```diff
@@ -119,7 +126,9 @@ export const bootstrapTestInstance: any = async () => {
 
     const connection = await app.resolve<Connection>(getConnectionToken());
     const tables = connection.entityMetadatas
-        .filter((e) => e.tableName !== 'issuer_blockchain_properties' && e.tableName !== 'issuer_signer')
+        .filter(
+            (e) => e.tableName !== 'issuer_blockchain_properties' && e.tableName !== 'issuer_signer'
```
Comment (soanvig) [inline packages/origin-247-certificate/test/setup.ts:130]: I think this doesn't apply anymore and needs to use new table name, right?

**Review by soanvig — CHANGES_REQUESTED** [review]

File: packages/origin-247-certificate/test/setup.ts (hunk 8 above)
Comment (patryk-kochanski) [reply packages/origin-247-certificate/test/setup.ts:130]: As long as we still have issuer-api dependency, it applies.

File: packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+import { providers, Signer, Wallet } from 'ethers';
+import { getProviderWithFallback } from '@energyweb/utils-general';
+
+@Injectable()
+export class BlockchainPropertiesService {
+    private primaryRPC: string;
+    private issuerPrivateKey: string;
+
+    constructor(
+        private configService: ConfigService,
```
Comment (JonaszJestem) [inline packages/origin-247-certificate/src/onchain-certificate/blockchain-properties.service.ts:14]: you can type `ConfigService<SomeConfigType>` so you can have to type `<string>` when getting specific variables, but that's just a note

File: packages/origin-247-certificate/test/setup.ts (hunk 8 above)
Comment (soanvig) [reply packages/origin-247-certificate/test/setup.ts:130]: Oh right, just remember to remove it later

**Review by soanvig — APPROVED** [review]

**Pull request merged**